stglib/vec/cdf2nc.py

```diff
@@ -396,3 +357,161 @@ def dist_to_boundary(ds):
 
     return ds
+
+
+def reshape(ds):
+
+    # find times of first sample in each burst
+    t = ds["time"][ds["sample"].values == 1]
+    # get corresponding burst number
+    b_num = ds["Burst"][ds["sample"].values == 1]
+
+    for i in np.arange(0, len(t)):
+        t2, samp = np.meshgrid(
+            t[i],
+            ds["sample"].sel(
+                time=slice(
+                    t[i],
+                    ds.time[ds["Burst"] == b_num[i]].max(),
+                )
+            ),
+        )
+
+        if i == 0:
+            s = np.array(samp.transpose().flatten())
+            t3 = np.array(t2.transpose().flatten())
+        else:
+            s = np.append(s, samp.transpose().flatten())
+            t3 = np.append(t3, t2.transpose().flatten())
+
+    ind = pd.MultiIndex.from_arrays((t3, s), names=("new_time", "new_sample"))
+
+    ds = ds.sel(time=slice(t[0], ds["time"][-1])).assign(time=ind).unstack("time")
+
+    ds = ds.drop("sample").rename({"new_time": "time", "new_sample": "sample"})
+
+    return ds
+
+
+def combine_vars(ds):
+    """Combines beam variables (cor, amp, snr) and raw (beam or xyz) velocities into matrix to limit number of variables"""
+
+    ds["AGC_1202"] = (ds["AMP1"] + ds["AMP2"] + ds["AMP3"]) / 3
+
+    veldim = ds.attrs["VECCoordinateSystem"].lower()
+
+    if veldim == "xyz":
+        veldim = "inst"
+
+    ds["vel"] = xr.DataArray([ds.VEL1, ds.VEL2, ds.VEL3], dims=[veldim, "time"])
+    ds["vel"].attrs.update(
+        {
+            "units": "m s-1",
+            "long_name": f"Velocity, {veldim} coordinate system",
+        }
+    )
+
+    ds["cor"] = xr.DataArray([ds.COR1, ds.COR2, ds.COR3], dims=["beam", "time"]).astype(
+        "int32"
+    )
+    ds["amp"] = xr.DataArray([ds.AMP1, ds.AMP2, ds.AMP3], dims=["beam", "time"]).astype(
+        "int32"
+    )
+    ds["snr"] = xr.DataArray([ds.SNR1, ds.SNR2, ds.SNR3], dims=["beam", "time"])
+
+    return ds
+
+
+def time_encoding(ds):
+    """ensure we don't set dtypes uint for CF compliance"""
+    if "units" in ds["time"].encoding:
+        ds["time"].encoding.pop("units")
+
+    # use time step to select time encoding
+    tstep = ds["time"][1] - ds["time"][0]
+
+    if tstep < np.timedelta64(1, "m"):
+
+        histtext = f"make time encoding to dtype double because tstep {tstep} seconds is < 1 minute, round to milliseconds first"
+        ds = utils.insert_history(ds, histtext)
+
+        # round time to milliseconds first
+        ds["time"] = ds["time"].dt.round("ms")
+        ds["time"].encoding["dtype"] = "double"
+
+    else:
+        histtext = f"make time encoding int because tstep {tstep} seconds is >= 1 minute, round time to seconds first"
+        ds = utils.insert_history(ds, histtext)
+
+        # round time to seconds if time interval >= 1 minute
+        ds["time"] = ds["time"].dt.round("s")
+
+        # check time to make sure it fits in int32, assume seconds for time units
+        utils.check_time_fits_in_int32(ds, "time")
```

Comment:
Call this higher up. If it fits, no problem; don't do any rounding. If it doesn't, then cast to double.